File: Logic/AuthorDM.cs
using EntityModels;
using Infrastructure.Data;
using Infrastructure.Logic;
using Shared.Interfaces;
using Shared.Services;
using System.Collections.Generic;
using ViewModels;

namespace Logic
{
    public class AuthorDM : BaseDM, IAuthorDM
    {
        public AuthorDM(IFactory factory) : base(factory)
        {

        }

        public IEnumerable<AuthorVM> Get(string term)
        {
            using (var authorRepo = Factory.GetService<IAuthorRepo>())
            {
                var authorsEM = authorRepo.Get(term);
                var authorsVM = MapperService.Map<IEnumerable<AuthorVM>>(authorsEM);
                return authorsVM;
            }
        }

        public DataTableResponseVM<AuthorVM> GetList(DataTableRequestVM model)
        {
            var dataTableEM = MapperService.Map<DataTableRequestEM>(model);

            using (var authorRepo = Factory.GetService<IAuthorRepo>())
            {
                var responseEM = authorRepo.GetList(dataTableEM);
                var responseVM = MapperService.Map<DataTableResponseVM<AuthorVM>>(responseEM);
                responseVM.draw = model.Draw;

                return responseVM;
            }           
        }

        public AuthorVM Get(long? id)
        {
            using (var authorRepo = Factory.GetService<IAuthorRepo>())
            {
                if (id.HasValue)
                {
                    var authorEM = authorRepo.Get(id.Value);
                    var authorVM = MapperService.Map<AuthorVM>(authorEM);
                    return authorVM;
                }
                return null;
            }
        }

        public void Delete(long id)
        {
            using (var authorRepo = Factory.GetService<IAuthorRepo>())
            {
                authorRepo.Delete(id);
            }
        }

        public void Update(AuthorVM model)
        {
            var author = MapperService.Map<AuthorEM>(model);

            using (var authorRepo = Factory.GetService<IAuthorRepo>())
            {
                authorRepo.Update(author);
            }
        }

        public void Create(AuthorVM model)
        {
            var author = MapperService.Map<AuthorEM>(model);

            using (var authorRepo = Factory.GetService<IAuthorRepo>())
            {
                authorRepo.Create(author);
            }
        }
    }
}

File: LogicVB/AuthorDM.vb
Imports EntityModels
Imports Infrastructure.Data
Imports Infrastructure.Logic
Imports [Shared].Interfaces
Imports [Shared].Services
Imports ViewModels

Public Class AuthorDM
        Inherits BaseDM
        Implements IAuthorDM

        Public Sub New(factory As IFactory)
            MyBase.New(factory)
        End Sub

        Public Sub Delete(id As Long) Implements IAuthorDM.Delete
            Using authorRepo = Factory.GetService(Of IAuthorRepo)()
                authorRepo.Delete(id)
            End Using
        End Sub

        Public Sub Update(model As AuthorVM) Implements IAuthorDM.Update
            Dim author = MapperService.Map(Of AuthorEM)(model)

            Using authorRepo = Factory.GetService(Of IAuthorRepo)()
                authorRepo.Update(author)
            End Using
        End Sub

        Public Sub Create(model As AuthorVM) Implements IAuthorDM.Create
            Dim author = MapperService.Map(Of AuthorEM)(model)

            Using authorRepo = Factory.GetService(Of IAuthorRepo)()
                authorRepo.Create(author)
            End Using

        End Sub

        Public Function [Get](term As String) As IEnumerable(Of AuthorVM) Implements IAuthorDM.Get
            Using authorRepo = Factory.GetService(Of IAuthorRepo)()
                Dim authorsEM = authorRepo.Get(term)
                Dim authorsVM = MapperService.Map(Of IEnumerable(Of AuthorVM))(authorsEM)
                Return authorsVM
            End Using
        End Function

        Public Function [Get](id As Long?) As AuthorVM Implements IAuthorDM.Get
            Using authorRepo = Factory.GetService(Of IAuthorRepo)()
                If (id.HasValue) Then
                    Dim AuthorEM = authorRepo.Get(id.Value)
                    Dim AuthorVM = MapperService.Map(Of AuthorVM)(AuthorEM)
                    Return AuthorVM
                End If
                Return Nothing
            End Using
        End Function

    Public Function GetList(model As DataTableRequestVM) As DataTableResponseVM(Of AuthorVM) Implements IAuthorDM.GetList
        Dim dataTableEM = MapperService.Map(Of DataTableRequestEM)(model)

        Using authorRepo = Factory.GetService(Of IAuthorRepo)()
            Dim responseEM = authorRepo.GetList(dataTableEM)
            Dim responseVM = MapperService.Map(Of DataTableResponseVM(Of AuthorVM))(responseEM)
            responseVM.draw = model.Draw

            Return responseVM
        End Using

    End Function
End Class